[guards][cpp-guards] Optimize NN module getattr guards

anijain2305 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124975
* #125202
* #125324
* __->__ #124522
* #125421
* #125439

Improves the guard overhead of MobileBert model with nn module guards from 92000 units to 20000 units.


cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng

anijain2305 committed:
[wip][guards][cpp-guards] Optimize NN module getattr guards

[ghstack-poisoned]

anijain2305 committed:
Update on "[wip][guards][cpp-guards] Optimize NN module getattr guards"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

anijain2305 committed:
Update on "[wip][guards][cpp-guards] Optimize NN module getattr guards"


Improves the guard overhead of MobileBert model with nn module guards from 92000 units to 34000 units.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

anijain2305 committed:
Update on "[guards][cpp-guards] Optimize NN module getattr guards"


Improves the guard overhead of MobileBert model with nn module guards from 92000 units to 34000 units.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

anijain2305 committed:
Update on "[guards][cpp-guards] Optimize NN module getattr guards"


Improves the guard overhead of MobileBert model with nn module guards from 92000 units to 34000 units.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

anijain2305 committed:
Update on "[guards][cpp-guards] Optimize NN module getattr guards"


Improves the guard overhead of MobileBert model with nn module guards from 92000 units to 34000 units.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

anijain2305 committed:
Update on "[guards][cpp-guards] Optimize NN module getattr guards"


Improves the guard overhead of MobileBert model with nn module guards from 92000 units to 38000 units.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

anijain2305 committed:
Update on "[guards][cpp-guards] Optimize NN module getattr guards"


Improves the guard overhead of MobileBert model with nn module guards from 92000 units to 38000 units.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

anijain2305 committed:
Update on "[guards][cpp-guards] Optimize NN module getattr guards"


Improves the guard overhead of MobileBert model with nn module guards from 92000 units to 20000 units.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

anijain2305 committed:
Update on "[guards][cpp-guards] Optimize NN module getattr guards"


Improves the guard overhead of MobileBert model with nn module guards from 92000 units to 20000 units.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

anijain2305 committed:
Update on "[guards][cpp-guards] Optimize NN module getattr guards"


Improves the guard overhead of MobileBert model with nn module guards from 92000 units to 20000 units.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

anijain2305 committed:
Update on "[guards][cpp-guards] Optimize NN module getattr guards"


Improves the guard overhead of MobileBert model with nn module guards from 92000 units to 20000 units.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

anijain2305 committed:
Update on "[guards][cpp-guards] Optimize NN module getattr guards"


Improves the guard overhead of MobileBert model with nn module guards from 92000 units to 20000 units.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

anijain2305 committed:
Update on "[guards][cpp-guards] Optimize NN module getattr guards"


Improves the guard overhead of MobileBert model with nn module guards from 92000 units to 20000 units.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

anijain2305 committed:
Update on "[guards][cpp-guards] Optimize NN module getattr guards"


Improves the guard overhead of MobileBert model with nn module guards from 92000 units to 20000 units.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

jansel reviewed (changes requested):
Add tests?

anijain2305 committed:
Update on "[guards][cpp-guards] Optimize NN module getattr guards"


Improves the guard overhead of MobileBert model with nn module guards from 92000 units to 20000 units.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

anijain2305 commented:
@jansel I added guards_nn_module to True to a few tests, which tests the added codebase more thoroughly. Actually, the tests today also test this PR partially because of `training` flag.

anijain2305 committed:
Update on "[guards][cpp-guards] Optimize NN module getattr guards"


Improves the guard overhead of MobileBert model with nn module guards from 92000 units to 20000 units.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

jansel commented on torch/_dynamo/guards.py:
```diff
@@ -526,6 +541,153 @@ def guard_on_dict_keys_and_order(self, value, guard):
+            mgr, key, source_name, base_example_value, example_value, guard_manager_enum
+        ):
+            if isinstance(mgr, DictGuardManager):
+                # Case where the user code relies on key order, e.g.,
```
Are there tests covering this case?  (Where we change the order)

anijain2305 replied:
There were none for NNModuleVariableTracker. There are for UnSpecializedNNModuleVariable.

Good catch. Added and made the fixes as well.

jansel commented on torch/_dynamo/guards.py:
```diff
@@ -526,6 +541,153 @@ def guard_on_dict_keys_and_order(self, value, guard):
+        for x in inspect.getmro(base_example_value.__class__):
+            all_class_attribute_names.update(x.__dict__.keys())
+
+        if attr_name in all_class_attribute_names:
```
class attributes are shadowed by module dict attributes... I think the order of these are wrong.  add some tests for shadowing.

anijain2305 replied:
You are right .. but there is a corner case with submodules, where class attr take precedence. This is the example

~~~
import torch
import torch._dynamo.testing

torch._dynamo.config.guard_nn_modules = True


class Mod(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.a = 3

    def forward(self,x, c=4):
        return x * c

    def linear(self, x):
        return x

    # b is class attribute and this is a bug that jansel pointed out
    # mod.b overshadows Mod.b
    def b(self, x):
        assert False
        return x

class MyMod(Mod):
    def __init__(self):
        super().__init__()
        # IF THIS LINE ACTUALLY CHAGES LINEAR, IT WOULD FAIL WITH SHAPE MISMATCH ERROR
        self.linear = torch.nn.Linear(11, 11)
        self.a = 2
        self.b = 2


    def forward(self, x, c=None):
        return self.linear(x) * self.a * self.b


mod = MyMod()
mod(torch.ones(3, 3))

# Instance attributes overwrite the instance attributes of the parent class.
assert mod.a == 2

# ANOMALY - mod.linear is not the parameter
assert mod.linear.__code__ is Mod.linear.__code__

# jansel's valid comment - module dict attr overshadow the class attrs
# anijain2305 - This does not cause a bug because today we fallback to getattr,
# but we are losing perf opportunity here.
assert mod.b == 2

cnts = torch._dynamo.testing.CompileCounter()
opt_mod = torch.compile(mod, backend=cnts)
opt_mod(torch.ones(3, 3))
opt_mod(torch.ones(3, 3))

assert cnts.frame_count == 1

~~~


The discrepancy is for `mod.linear`. Parent class has a function named `linear`. When the derived class instance calls self.linear, it calls the parent class function.

This is a special case. Apart from that everything else conforms to the general rule - "class attributes are shadowed by the mod dict attributes".

I also think this is a bug in `__setattr__` of nn.Module. For example, I am unable to assign a `self.param` if the parent class already has a function named `param`. If you think, this is just a bug, I can send a PR for nn modules.

jansel reviewed (changes requested):
add some more tests for this

anijain2305 committed:
Update on "[guards][cpp-guards] Optimize NN module getattr guards"


Improves the guard overhead of MobileBert model with nn module guards from 92000 units to 20000 units.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

anijain2305 committed:
Update on "[guards][cpp-guards] Optimize NN module getattr guards"


Improves the guard overhead of MobileBert model with nn module guards from 92000 units to 20000 units.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

anijain2305 committed:
Update on "[guards][cpp-guards] Optimize NN module getattr guards"


Improves the guard overhead of MobileBert model with nn module guards from 92000 units to 20000 units.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

anijain2305 committed:
Update on "[guards][cpp-guards] Optimize NN module getattr guards"


Improves the guard overhead of MobileBert model with nn module guards from 92000 units to 20000 units.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

anijain2305 committed:
Update on "[guards][cpp-guards] Optimize NN module getattr guards"


Improves the guard overhead of MobileBert model with nn module guards from 92000 units to 20000 units.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

anijain2305 committed:
Update on "[guards][cpp-guards] Optimize NN module getattr guards"


Improves the guard overhead of MobileBert model with nn module guards from 92000 units to 20000 units.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

anijain2305 committed:
Update on "[guards][cpp-guards] Optimize NN module getattr guards"


Improves the guard overhead of MobileBert model with nn module guards from 92000 units to 20000 units.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

anijain2305 committed:
Update on "[guards][cpp-guards] Optimize NN module getattr guards"


Improves the guard overhead of MobileBert model with nn module guards from 92000 units to 20000 units.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[guards][cpp-guards] Optimize NN module getattr guards"


Improves the guard overhead of MobileBert model with nn module guards from 92000 units to 20000 units.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[guards][cpp-guards] Optimize NN module getattr guards"


Improves the guard overhead of MobileBert model with nn module guards from 92000 units to 20000 units.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng

[ghstack-poisoned]

anijain2305 commented:
@jansel just a gentle ping in case this slipped in your notifications.
